In [48]:
import pandas as pd
import numpy as np

In [49]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [50]:
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

### Q1. Pandas version
What's the version of Pandas that you installed?

In [51]:
 pd.__version__

'2.3.1'

### Q2. Records count
How many records are in the dataset?

In [52]:
df.shape

(9704, 11)

In [53]:
print(f"In the dataset there are {df.shape[0]} records")

In the dataset there are 9704 records


### Q3. Fuel types

How many fuel types are presented in the dataset?

In [55]:
df['fuel_type'].nunique()
print(f"In the dataset there are {df['fuel_type'].nunique()} fuel types")

In the dataset there are 2 fuel types


### Q4. Missing values

How many columns in the dataset have missing values?

In [56]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [57]:
cols = df.columns
null_count = []


for i in cols:
    if df[i].isnull().sum()>0:
        null_count.append(1)
    n_null_col = sum(null_count)
        

print(f"The number of columns with missing values is {n_null_col}")

The number of columns with missing values is 4


### Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [60]:
asian_cars_fuel_eff =df[df['origin']=='Asia']['fuel_efficiency_mpg'].max().round()
print(f"The maximum fuel efficiency of cars from Asia is {asian_cars_fuel_eff}")

The maximum fuel efficiency of cars from Asia is 24.0


In [59]:
df.groupby('origin')['fuel_efficiency_mpg'].max().round()

origin
Asia      24.0
Europe    26.0
USA       25.0
Name: fuel_efficiency_mpg, dtype: float64

### Q6. Median value of horsepower

1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

In [61]:
# 1. Find the median value of horsepower column in the dataset.
df['horsepower'].median()

np.float64(149.0)

In [62]:
# 2. Next, calculate the most frequent value of the same horsepower column.
df['horsepower'].value_counts()

horsepower
152.0    142
145.0    141
151.0    134
148.0    130
141.0    130
        ... 
40.0       1
57.0       1
245.0      1
252.0      1
61.0       1
Name: count, Length: 192, dtype: int64

In [63]:
df['horsepower'].value_counts().max()

np.int64(142)

In [64]:
# 3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.

#checking the missing values
df['horsepower'].isnull().sum()

np.int64(708)

In [65]:
max_count = df['horsepower'].value_counts().max()

#fills all the missing values on the column horsepower by the most frequent value
df['horsepower'] = df['horsepower'].fillna(max_count)

In [66]:
#check that the null values were replaced
df['horsepower'].isnull().sum()

np.int64(0)

In [67]:
# 4. Now, calculate the median value of horsepower once again.
df['horsepower'].median()

np.float64(146.0)

### The median value has decreased from 149 to 146 after replacing the null values

### Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values 
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Compute the inverse of XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

In [72]:
# 1. Select all the cars from Asia from the dataset.
df[df['origin']=='Asia']

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       142.0     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [73]:
# 2. Select only columns vehicle_weight and model_year
df[df['origin']=='Asia'][['vehicle_weight','model_year']]

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [74]:
# 3. Select the first 7 values
df_asian_cars = df.copy()
df_asian_cars = df[df['origin']=='Asia'][['vehicle_weight','model_year']].head(7)
df_asian_cars

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [75]:
# 4. Get the underlying NumPy array. Let's call it X.
X = df_asian_cars.values
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [76]:
# 5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
XT =X.T
XT

array([[2714.21930965, 2783.86897424, 3582.68736772, 2231.8081416 ,
        2659.43145076, 2844.22753389, 3761.99403819],
       [2016.        , 2010.        , 2007.        , 2011.        ,
        2016.        , 2014.        , 2019.        ]])

In [77]:
XT.shape

(2, 7)

In [78]:
XTX = XT.dot(X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [79]:
XTX.shape

(2, 2)

In [80]:
# 6. Compute the inverse of XTX.
inv_XTX = np.linalg.inv(XTX)
inv_XTX


array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [81]:
I_XTX.shape

(2, 2)

In [82]:
# 7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

In [85]:
# 8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
w= inv_XTX_multip_XT_multip_y = inv_XTX.dot(XT).dot(y)



In [87]:
w

array([0.01386421, 0.5049067 ])

In [91]:
# 9. What's the sum of all the elements of the result?
print(f"The sum of all the elements of the result is {w.sum().round(3)}")

The sum of all the elements of the result is 0.519
